leif/parse_slo_rules.go

```diff
@@ -0,0 +1,201 @@
+		return err
+	}
+
+	value, isNumber := tempInterface.(float64)
+
+	if isNumber {
+		*stringOrInt = duration(time.Duration(int64(value) * int64(oneSec.Nanoseconds())))
+		return err
+	}
+
+	return errors.New("Invalid duration format")
+}
+
+type priority string
+type issueType string
+
+type sloRuleJSON struct {
+	AppliesToJSON          AppliesToJSON          `json:"appliesTo"`
+	ComplianceSettingsJSON ComplianceSettingsJSON `json:"complianceSettings"`
+}
+
+// returns a new SLORuleJSON with the defaults applied
+// except for the responders default which requires knowing if it was partially assigned
+func newSLORuleJSON() *sloRuleJSON {
+	return &sloRuleJSON{
+		AppliesToJSON:          AppliesToJSON{Issues: true, PRs: false},
+		ComplianceSettingsJSON: ComplianceSettingsJSON{RequiresAssignee: false},
+	}
+}
+
+func (rule *sloRuleJSON) addToGitHubLabels(prepend string, property string) {
+	if len(property) > 0 {
+		rule.AppliesToJSON.GitHubLabelsRaw = append(rule.AppliesToJSON.GitHubLabelsRaw, prepend+property)
+	}
+}
+
+func (rule *sloRuleJSON) applyResponderDefault() { //if not defined
+	if rule.ComplianceSettingsJSON.RespondersJSON.OwnersRaw == nil &&
+		len(rule.ComplianceSettingsJSON.RespondersJSON.Contributors) < 1 &&
+		rule.ComplianceSettingsJSON.RespondersJSON.Users == nil {
+		rule.ComplianceSettingsJSON.RespondersJSON.Contributors = "WRITE"
+	}
+}
+
+type AppliesToJSON struct {
+	GitHubLabelsRaw         stringOrArray `json:"gitHubLabels"`
+	ExcludedGitHubLabelsRaw stringOrArray `json:"excludedGitHubLabels"`
+	Priority                string        `json:"priority"`
+	IssueType               string        `json:"issueType"`
+	Issues                  bool          `json:"issues"`
+	PRs                     bool          `json:"prs"`
+}
+
+type ComplianceSettingsJSON struct {
+	ResponseTime     duration       `json:"responseTime"`
+	ResolutionTime   duration       `json:"resolutionTime"`
+	RequiresAssignee bool           `json:"requiresAssignee"`
+	RespondersJSON   RespondersJSON `json:"responders"`
+}
+
+type RespondersJSON struct {
+	OwnersRaw    stringOrArray `json:"owners"`
+	Contributors string        `json:"contributors"`
+	Users        []string      `json:"users"`
+}
+
+func ParseDurationWithDays(duration string) (time.Duration, error) {
+	if strings.Contains(duration, ".") {
+		return 0, errors.New("Duration should not contain fractions")
+	}
+	str := dayReg.ReplaceAllStringFunc(duration, func(s string) string {
+		days, _ := strconv.Atoi(s[:len(s)-1])
+		s = strconv.Itoa(days*24) + "h"
+		return s
+	})
+	return time.ParseDuration(str)
+}
+
+func parseSLORule(rawRule *json.RawMessage) (*SLORule, error) {
+	jsonRule := newSLORuleJSON() //apply defaults
+
+	err := json.Unmarshal(*rawRule, &jsonRule) //convert possible strings to arrays
+	if err != nil {
+		return nil, err
+	}
+
+	jsonRule.addToGitHubLabels("priority: ", jsonRule.AppliesToJSON.Priority)
+	jsonRule.addToGitHubLabels("type: ", jsonRule.AppliesToJSON.IssueType)
```

Comment:
I'm not sure I understand exactly what these two lines are doing. Can you clarify?

Reply:
This is to cover the case of a priority or issue type being set outside of `gitHubLabels` in the config. For example, in the JSON schema we allow people to specify:

```
{
   "appliesTo" : {
      "gitHubLabels": ["label1", "label2"],
      "priority": "P1",
      "issueType": "bug"
   }
// etc
}
```
So these lines add those to the labels (since leif would not have a priority field itself under `AppliesTo`)

Reply:
Okay this makes sense. I would, however, add a check like:

```go
if len(jssonRule.AppliesToJSON.Priority) > 0 {
    jsonRule.addToGitHubLabels("priority: ", jsonRule.AppliesToJSON.Priority)
}
```

To indicate that this is explicitly catching the case where the Priority is Specified in the rule.